source/actions3/actions/AMControlCalibrateAction.cpp

```diff
@@ -0,0 +1,121 @@
+#include "util/AMErrorMonitor.h"
+
+AMControlCalibrateAction::AMControlCalibrateAction(AMControlCalibrateActionInfo *info, AMControl *control, QObject *parent) :
+	AMAction3(info, parent)
+{
+	// Initialize class variables.
+
+	control_ = 0;
+
+	// Current settings.
+
+	setControl(control);
+}
+
+AMControlCalibrateAction::AMControlCalibrateAction(const AMControlCalibrateAction &original) :
+	AMAction3(original)
+{
+	setControl(original.control());
+}
+
+AMControlCalibrateAction::~AMControlCalibrateAction()
+{
+
+}
+
+void AMControlCalibrateAction::setControl(AMControl *control)
+{
+	if (control)
+		control_ = control;
+	else
+		control_ = AMBeamline::bl()->exposedControlByInfo(controlCalibrateInfo()->controlInfo());
+}
+
+void AMControlCalibrateAction::onCalibrateStarted()
+{
+	// Set action as started.
+
+	setStarted();
+
+	// Start progress timer updates.
+
+	connect( &progressTimer_, SIGNAL(timeout()), this, SLOT(onProgressTick()) );
+	progressTimer_.start(250);
+}
+
+void AMControlCalibrateAction::onCalibrateFailed()
+{
+	// Disconnect from the control.
+
+	disconnect(control_, 0, this, 0);
+
+	// End progress timer updates.
+
+	progressTimer_.stop();
+	disconnect(&progressTimer_, 0, this, 0);
+
+	// Create failure message and set action as failed.
+
+	QString fundamentalFailureMessage = QString("There was an error calibrating the control '%1' such that %2 becomes %3.").arg(control_->name()).arg(controlCalibrateInfo()->oldValue()).arg(controlCalibrateInfo()->newValue());
+	setFailed(fundamentalFailureMessage);
+}
+
+void AMControlCalibrateAction::onCalibrateSucceeded()
+{
+	// Disconnect from the control.
+
+	disconnect( control_, 0, this, 0 );
+
+	// End progress timer updates.
+
+	progressTimer_.stop();
+	disconnect( &progressTimer_, 0, this, 0 );
+	setProgress(100, 100);
+
+	// Set action as succeeded.
+
+	setSucceeded();
+}
+
+void AMControlCalibrateAction::startImplementation()
+{
+	// Must have a control, and it must support calibration.
+
+	if(!control_) {
+		QString fundamentalFailureMessage = QString("There was an error calibrating the control '%1', because the control was not found.").arg(controlCalibrateInfo()->controlInfo().name());
+		AMErrorMon::alert(this, AMCONTROLCALIBRATEACTION_CONTROL_NOT_VALID, QString("%1. Please report this problem to the Acquaman developers.").arg(fundamentalFailureMessage));
+		setFailed(fundamentalFailureMessage);
+		return;
+	}
+
+	if(!control_->canCalibrate()) {
+		QString fundamentalFailureMessage = QString("There was an error calibrating the control '%1', because the control reported it cannot calibrate.").arg(control_->name());
+		AMErrorMon::alert(this, AMCONTROLCALIBRATEACTION_CONTROL_CANT_CALIBRATE, QString("%1. Please report this problem to the beamline staff.").arg(fundamentalFailureMessage));
+		setFailed(fundamentalFailureMessage);
+		return;
+	}
+
+	// Check that the calibration values are valid.
+
+	if (controlCalibrateInfo()->oldValue() == controlCalibrateInfo()->newValue()) {
```

Comment:
Ah, I see.  Yeah, I think if that is the case, then making the values AMNumbers would be better (or some other kind of test, like hasValidValues() or something).

As it relates to my original comment, it might be preferable to have the action automatically succeed if the values are actually the same.